walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,162 @@
 
 package com.google.sps;
 
+import com.google.common.collect.Iterables; 
+import com.google.common.collect.Sets;
+import java.lang.Math;
+import java.util.Arrays;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
+  /*
+   * Function to return a collection of time ranges where people are available to meet
+   * 
+   * @param events    A collection of events which has attendees, our results should avoid these events
+   * @param request   A specification on the minimum duration, mandatory and optional attendees
+   * 
+   * @return          A collection of time ranges where attendees are available
+   * 
+  */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if (request.getAttendees().size() <= 0 && request.getOptionalAttendees().size() <= 0) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }  
+    
+    HashSet<String> mandatoryAttendees = new HashSet<String>(request.getAttendees());
+    HashSet<String> optionalAttendees = new HashSet<String>(request.getOptionalAttendees());
+    HashSet<String> allAttendees = Sets.newHashSet(Iterables.concat(mandatoryAttendees, optionalAttendees));
+      
+    String endOfDayTitle = "END_OF_DAY";
+    String startOfDayTitle = "START_OF_DAY";
+    Event endOfDay = new Event(endOfDayTitle, TimeRange.fromStartDuration(TimeRange.END_OF_DAY + 1, 0), allAttendees);
+    Event startOfDay = new Event(startOfDayTitle, TimeRange.fromStartDuration(TimeRange.START_OF_DAY, 0), allAttendees);
+
+    ArrayList<Event> allEvents = new ArrayList<Event>(events);
+
+    allEvents.add(endOfDay);
+    allEvents.add(startOfDay);
+      
+    allEvents.sort(Event.EVENT_COMPARATOR);
+    
+    ArrayList<TimeRange> available = availableTimesFromSortedEvents(mandatoryAttendees, allEvents, request);
+
+    if (optionalAttendees.size() == 0) {
+      return available;
+    }
+
+    ArrayList<TimeRange> availableWithOptional = availableTimesFromSortedEvents(optionalAttendees, allEvents, request);
+    
+    if (available.size() == 0) {
+      return availableWithOptional;
```

Comment:
I just felt it to be more natural if when there's no available time for the mandatory attendees to then proceed to accommodate the optionals instead of scraping the meeting request. An example is in the noMandatoryTimeslotAvailable unit test which I wrote for this specific case.

Reply:
Ah ok, that seems reasonable. I'll point out that Google Calendar and Microsoft Outlook would probably return empty in this scenario. But you are free to innovate, this is your scheduling algo.